### PR TITLE
feat: Add notebook sample for Building a Conversational Search Agent with Reasoning Engine and RAG

### PR DESCRIPTION
**Pull request by koverholt**
# Description

Adds a sample notebook for "Building a Conversational Search Agent with Reasoning Engine and RAG on Vertex AI Search".

- [x] Follow the [`CONTRIBUTING` Guide](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/CONTRIBUTING.md).
- [x] You are listed as the author in your notebook or README file.
  - [x] Your account is listed in [`CODEOWNERS`](https://github.com/GoogleCloudPlatform/generative-ai/blob/main/.github/CODEOWNERS) for the file(s).
- [x] Make your Pull Request title in the <https://www.conventionalcommits.org/> specification.
- [x] Ensure the tests and linter pass (Run `nox -s format` from the repository root to format).
- [x] Appropriate docs were updated (if necessary)

### CODE REVIEW

File: gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb [hunk 1]
```diff
@@ -0,0 +1,823 @@
...
...AhKQgAQkIAEJSEACEpCABCTQCAIKjY1oBishAQlIQAISkIAEJCABCUhAAhKQgAQkIIF2E1BobHf7WXsJSEACEpCABCQgAQlIQAISkIAEJCABCTSCgEJjI5rBSkhAAhKQgAQkIAEJSEACEpCABCQgAQlIoN0EFBrb3X7WXgISkIAEJCBaqZuwAAAC4UlEQVQBCUhAAhKQgAQkIAEJSEACjSCg0NiIZrASEpCABCQgAQlIQAISkIAEJCABCUhAAhJoNwGFxna3n7WXgAQkIAEJSEACEpCABCQgAQlIQAISkEAjCCg0NqIZrIQEJCABCUhAAhKQgAQkIAEJSEACEpCABNpNQKGx3e1n7SUgAQlIQAISkIAEJCABCUhAAhKQgAQk0AgCCo2NaAYrIQEJSEACEpCABCQgAQlIQAISkIAEJCCBdhNQaGx3+1l7CUhAAhKQgAQkIAEJSEACEpCABCQgAQk0goBCYyOawUpIQAISkIAEJCABCUhAAhKQgAQkIAEJSKDdBBQa291+1l4CEpCABCQgAQlIQAISkIAEJCABCUhAAo0goNDYiGawEhKQgAQkIAEJSEACEpCABCQgAQlIQAISaDcBhcZ2t5+1l4AEJCABCUhAAhKQgAQkIAEJSEACEpBAIwgoNDaiGayEBCQgAQlIQAISkIAEJCABCUhAAhKQgATaTUChsd3tZ+0lIAEJSEACEpCABCQgAQlIQAISkIAEJNAIAgqNjWgGKyEBCUhAAhKQgAQkIAEJSEACEpCABCQggXYTUGhsd/tZewlIQAISkIAEJCABCUhAAhKQgAQkIAEJNIKAQmMjmsFKSEACEpCABCQgAQlIQAISkIAEJCABCUig3QQUGtvdftZeAhKQgAQkIAEJSEACEpCABCQgAQlIQAKNIKDQ2IhmsBISkIAEJCABCUhAAhKQgAQkIAEJSEACEmg3AYXGdreftZeABCQgAQlIQAISkIAEJCABCUhAAhKQQCMIKDQ2ohmshAQkIAEJSEACEpCABCQgAQlIQAISkIAE2k1AobHd7WftJSABCUhAAhKQgAQkIAEJSEACEpCABCTQCAIKjY1oBishAQlIQAISkIAEJCABCUhAAhKQgAQkIIF2E1BobHf7WXsJSEACEpCABCQgAQlIQAISkIAEJCABCTSCwP8FN7nmt/SlyMcAAAAASUVORK5CYII=\">"
```
Comment (holtskinner) [inline gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb:338]: For this image, can you upload it to the GCS Bucket and link it in? These embedded images can make the notebooks slower to download and render.

https://console.cloud.google.com/storage/browser/github-repo;tab=objects?prefix=&forceOnObjectsSortingFiltering=false

File: gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb [hunk 2]
```diff
@@ -0,0 +1,823 @@
...
+    "        (\"user\", \"{input}\"),\n",
+    "        (\"placeholder\", \"{agent_scratchpad}\"),\n",
+    "    ]\n",
+    ")\n",
+    "\n",
+    "# Initialize session history\n",
+    "store = {}\n",
+    "\n",
+    "\n",
+    "def get_session_history(session_id: str):\n",
```
Comment (holtskinner) [inline gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb:449]: Can you add a return type for this method?

File: gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb [hunk 3]
```diff
@@ -0,0 +1,823 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 16,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Retrieve the project number associated with your project ID\n",
+    "client = resourcemanager_v3.ProjectsClient()\n",
+    "project = client.get_project(name=f\"projects/{PROJECT_ID}\")\n",
```
Comment (holtskinner) [inline gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb:609]: nit: There is a method on the client to construct a `projects/` path

https://cloud.google.com/python/docs/reference/cloudresourcemanager/latest/google.cloud.resourcemanager_v3.services.projects.ProjectsClient#google_cloud_resourcemanager_v3_services_projects_ProjectsClient_project_path

And a method to parse it. https://cloud.google.com/python/docs/reference/cloudresourcemanager/latest/google.cloud.resourcemanager_v3.services.projects.ProjectsClient#google_cloud_resourcemanager_v3_services_projects_ProjectsClient_parse_project_path

It's a little less hacky than constructing/splitting the strings manually.

**Review by holtskinner — APPROVED** [review]
Very cool! I'm glad that we have something to get good quality RAG from structured Data VAIS 😛

File: gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb (hunk 1 above)
Comment (koverholt) [reply gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb:338]: Makes sense, what about an `images/` folder within `gemini/reasoning-engine` to keep it closer to the content?

File: gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb (hunk 2 above)
Comment (koverholt) [reply gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb:444]: Done!

File: gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb (hunk 3 above)
Comment (koverholt) [reply gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb]: Thanks for the tip! I was not able to use the `project_path` or `parse_project_path` methods to go from a Project ID to a Project number. But I was able to swap to a different approach to get the project number that doesn't use splitting:

```python
from googleapiclient import discovery
service = discovery.build('cloudresourcemanager', 'v1')
request = service.projects().get(projectId=PROJECT_ID)
response = request.execute()
project_number = response['projectNumber']
project_number
```

File: gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb (hunk 1 above)
Comment (koverholt) [reply gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb]: Got it, thanks for the tip. I need write access to that bucket and I'll put the image there.

Comment (koverholt) [reply gemini/reasoning-engine/tutorial_vertex_ai_search_rag_agent.ipynb]: Done, thanks!

**Pull request merged**